crates/rrg/src/action/get_filesystem_timeline_tsk.rs

```diff
@@ -0,0 +1,582 @@
+            mountpoint: mount.path.to_path_buf(),
+        })
+    }
+}
+
+/// Given an absolute path and the raw device that contains it, computes the path
+/// embedded within the raw device to give TSK.
+///
+/// This path must start with `/`, and this function normalizes all path
+/// separators in the embedded path to use `/` for consistency.
+fn get_embedded_path(mount: &RawDeviceMount, absolute_path: &Path) -> PathBuf {
+    absolute_path
+        .strip_prefix(&mount.mountpoint)
+        .expect("mountpoint not prefix of root")
+        .iter()
+        .flat_map(|component| [OsStr::new("/"), component])
+        .collect::<OsString>()
+        .into()
+}
+
+/// Handles requests for the timeline action.
+pub fn handle<S>(session: &mut S, args: Args) -> crate::session::Result<()>
+where
+    S: crate::session::Session,
+{
+    use sha2::Digest as _;
+
+    let raw_fs: PathBuf;
+    let embedded_path;
+    if !args.raw_fs.as_os_str().is_empty() {
+        raw_fs = args.raw_fs;
+        embedded_path = args.root;
+    } else {
+        // Detect raw filesystem
+        let mounts = ospect::fs::mounts()
+            .and_then(|iter| iter.collect::<Result<Vec<_>, _>>())
+            .map_err(crate::session::Error::action)?;
+        let mount = get_raw_device(&mounts, &args.root).map_err(crate::session::Error::action)?;
+        embedded_path = get_embedded_path(&mount, &args.root);
+        // The raw fs path may be e.g. /dev/sda1 on Linux or \\?\C: on Windows.
+        raw_fs = mount.image_path;
+    };
+
+    let (tx, rx) = std::sync::mpsc::sync_channel(0);
+    let tsk_thread = std::thread::spawn(move || -> tsk::TskResult<()> {
+        let image = tsk::TskImage::open(&raw_fs)?;
+        let mut fs = tsk::TskFs::open(&image)?;
+        let dir = fs.open_dir(&embedded_path)?;
+        let root_bytes = embedded_path.as_os_str().as_encoded_bytes();
+        fs.walk_dir(dir.addr(), |file, path| {
+            match tx.send(serialize_entry(root_bytes, path, file)) {
+                Ok(()) => tsk::WalkDirCallbackResult::Continue,
+                // rx disconnected, no need to keep searching.
+                Err(_) => tsk::WalkDirCallbackResult::Stop,
+            }
+        })
+    });
+
+    // `entry_count` keeps track of the number of entries that are included in
+    // each batch. Each time the `entries` iterator (defined below) yields an
+    // entry, we increase the count (through `Iterator::inspect`). We read the
+    // of `entry_count` when we are about to send a batch and reset the counter
+    // so that we start from 0 for the next batch.
+    //
+    // Note that `entry_count` has to be a cell because it is mutably borrowed
+    // by the `entries` iterator but we still want to be able to also modify it
+    // when we process batches.
+    let entry_count = &std::cell::RefCell::new(0);
+
+    let result_iter = rx.into_iter().inspect(|_| *entry_count.borrow_mut() += 1);
```

Comment:
[TIL](https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html#impl-IntoIterator-for-Receiver%3CT%3E) :o